Enable custom test list

desmetbruno opened the pull request:
**Context**
As the list of test cases is built at compile time, if some test cases need to be disabled, custom NSPE binaries and a custom test executable need to be built.
This is the case when
- the test suite does not match the capabilities of the implementation
- the implementation and the test suite are incompatible (updated the test suite or implementation under development)

**Improvement**
Implement a mechanism to customize the list of test cases at runtime.
This mechanism defines the list of runable test cases among the list of test cases implemented by the test suite.
Using to this mechanism, generic NSPE binaries can be used for platforms with different capabilities.

**This PR**
Added 2 PAL functions:
pal_set_custom_test_list() - Sets the custom test list buffer
pal_is_test_enabled() - Tells if a test is enabled on platform

A platform can provide the custom test case list by calling pal_set_custom_test_list().
The Validation Abstraction Layer (VAL) calls pal_is_test_enabled() to check if a test case is enabled and eventually runs it.
All test cases are enabled if no custom list of test cases is provided.
Platforms that do not need to implement this mechanism do not have to implement these 2 PAL functions.

desmetbruno committed:
Enable custom test list

Added 2 PAL functions:
pal_set_custom_test_list() - Sets the custom test list buffer
pal_is_test_enabled() - Tells if a test is enabled on platform

The platform can provide the custom test list by calling
pal_set_custom_test_list().
The Validation Abstraction Layer (VAL) calls pal_is_test_enabled()
to check if a test is enabled and eventually runs it.
All tests are enabled if no custom test list is provided.

Signed-off-by: Bruno De Smet <bruno.de.smet@nxp.com>

jk-arm committed:
Update test_failure_analysis.md

cneveux commented on api-tests/CMakeLists.txt:
```diff
@@ -647,6 +648,8 @@ if(${CC312_LEGACY_DRIVER_API_ENABLED})
 endif()
 
 # Build PAL NSPE LIB
+list(APPEND PAL_SRC_C_NSPE
```
I'm not sure this is a right approach. PAL_SRC_C_NSPE is platform/targets/xxx/target.cmake list of files and the pal_weak.c is also platform/targets file.
According to the overall project, platform/targets files must be added in the PAL_SRC_C_NSPE list from the platform/targets/xxx/target.cmake.

In my mind, adding a "platform/targets/common/nspe/target.cmake" including this file will be better, but I think will be better to highlight this problem in the PR proposal to ask what is the best approach in order to fit ARM test suite implementation rule.

desmetbruno replied:
Good point.
I will push a fixup commit implementing this. And I will ask ARM to give their recommendation.

jk-arm committed:
provision added to skip a testcases from regression

Signed-off-by: jotman01 <jothikumar.mani@arm.com>

jk-arm committed:
giving option to use hardcoded/platform specific keys

Signed-off-by: jotman01 <jothikumar.mani@arm.com>

desmetbruno committed:
fixup! Enable custom test list

Signed-off-by: Bruno De Smet <bruno.de.smet@nxp.com>

desmetbruno committed:
Provide the list of available tests

Added 1 VAL function:
val_get_test_list() - Returns the IDs list of available tests

Signed-off-by: Bruno De Smet <bruno.de.smet@nxp.com>

jk-arm committed:
cmake related issue fixed github issue #344

Signed-off-by: jothikumar mani <jothikumar.mani@arm.com>

david-hazi-arm committed:
Add corstone-300 and corstone-310 platforms

* Add V8.1M arch support
* Fix watchdog enable
* Add cmsdk uart baudrate setup
* Fix ident in modified files (\t -> 4*whitespace)v

jk-arm committed:
Merge pull request #349 from david-hazi-arm/corstone_3x0

Add corstone-300 and corstone-310 platforms

jk-arm committed:
Merge branch 'main' into custom_test_list
